[macOS] Create a VPIO unit in a background thread

youennf opened the pull request:
#### 4f1a121947d3614fbc3edf5f372ec7c297c38074
<pre>
[macOS] Create a VPIO unit in a background thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=265017">https://bugs.webkit.org/show_bug.cgi?id=265017</a>
<a href="https://rdar.apple.com/118520768">rdar://118520768</a>

Reviewed by Jean-Yves Avenard.

A VPIO Unit may take a few seconds to be created.
This might block the GPU process main thread, which might block the WebProcess main thread.
The impact might be a page freeze.
To prevent this, we create the VPIO unit in a background queue and asynchronously start capturing.

To do so, CoreAudioCaptureSource is implementing RealtimeMediaSource::whenReady, where we prewarm the audio unit.
The impact is to delay the resolution of the getUserMedia promise.

To improve the handling of AudioUnit, we introduce StoredAudioUnit to make sure we always dispose a created audio unit.
We then refactor the code to use move-only audio units.

Covered by existing tests in macOS with MockAudioSharedUnit.
Manually tested as well.

* Source/WTF/wtf/NativePromise.h:
* Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.h:
(WebCore::BaseAudioSharedUnit::prewarmAudioUnitCreation):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::whenReady):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
(WebCore::CoreAudioSharedUnit::AudioUnitDeallocator::operator() const):
(WebCore::createAudioUnit):
(WebCore::CoreAudioSharedInternalUnit::create):
(WebCore::CoreAudioSharedInternalUnit::CoreAudioSharedInternalUnit):
(WebCore::CoreAudioSharedInternalUnit::~CoreAudioSharedInternalUnit):
(WebCore::CoreAudioSharedInternalUnit::initialize):
(WebCore::CoreAudioSharedInternalUnit::uninitialize):
(WebCore::CoreAudioSharedInternalUnit::start):
(WebCore::CoreAudioSharedInternalUnit::stop):
(WebCore::CoreAudioSharedInternalUnit::set):
(WebCore::CoreAudioSharedInternalUnit::get):
(WebCore::CoreAudioSharedInternalUnit::render):
(WebCore::CoreAudioSharedUnit::~CoreAudioSharedUnit):
(WebCore::CoreAudioSharedUnit::setStoredVPIOUnit):
(WebCore::CoreAudioSharedUnit::reconfigureAudioUnit):
(WebCore::CoreAudioSharedUnit::migrateToNewDefaultDevice):
(WebCore::CoreAudioSharedUnit::prewarmAudioUnitCreation):
(WebCore::CoreAudioSharedInternalUnit::storeVPIOUnitIfNeeded): Deleted.
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):

Canonical link: <a href="https://commits.webkit.org/271050@main">https://commits.webkit.org/271050@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb8f9925a924110cac1f11861559a3aabb64930b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27177 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29390 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24860 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3215 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24685 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27439 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4608 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23317 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4020 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4127 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30024 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23647 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24809 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24728 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30309 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26357 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4163 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2317 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28236 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5629 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33813 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6549 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4622 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7321 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4543 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,33 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
+        WorkQueue::create("AudioUnit creation")->dispatch([weakThis = WeakPtr { *this }, callback = WTFMove(callback)] () mutable {
+            callOnMainRunLoop([weakThis = WTFMove(weakThis), callback = WTFMove(callback), vpioUnitOrError = createAudioUnit(true)] () mutable {
+                if (!weakThis)
+                    return;
```
need to call the callback

jyavenard replied:
where is the captured callback running ?

jyavenard replied:
oh the completionHandler captured was moved in the line above, so it's unused

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
```diff
@@ -31,6 +31,7 @@
 #include "BaseAudioSharedUnit.h"
 #include "CAAudioStreamDescription.h"
 #include "Timer.h"
+#include <wtf/NativePromise.h>
```
wtf/Forward.h no need to include it. Only in the code file

jyavenard replied:
`#include <wtf/Forward.h>`
and in code #include wtf/NativePromise.h would likely provide faster compilation time

jyavenard commented on Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.h:
```diff
@@ -55,6 +55,7 @@ class BaseAudioSharedUnit : public RealtimeMediaSourceCenter::Observer {
     virtual bool isProducingData() const = 0;
 
     virtual void delaySamples(Seconds) { }
+    virtual void prewarmAudioUnitCreation(CompletionHandler<void()>&& callback) { callback(); };
```
this should return a `Ref<GenericNonExclusivePromise>`  to simplify the code later greatly.

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
```diff
@@ -343,6 +343,15 @@ void CoreAudioCaptureSource::setIsInBackground(bool value)
+#if PLATFORM(MAC)
+void CoreAudioCaptureSource::whenReady(CompletionHandler<void(CaptureSourceError&&)>&& callback)
+{
+    unit().prewarmAudioUnitCreation([callback = WTFMove(callback)] () mutable {
```
then you would do:
`unit().prewarmAudioUnitCreation()->whenSettled(RunLoop::current(), [callback = WTFMove(callback)]() mutable { callback(); });`

(may want to handle the case there's an error, it's ignored right now)

jyavenard replied:
From an offline discussing, you don't want to expose promise outside the CoreAudioUnit code, so I leave it at that

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,30 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
 }
-#endif
 
+void CoreAudioSharedUnit::prewarmAudioUnitCreation(CompletionHandler<void()>&& callback)
```
if prewarnAudioUnitCreation returned a GenericNonExclusivePromise you would do:

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,30 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
+    }
+
+    if (!m_audioUnitCreationWarmupPromise->isSettled()) {
+        m_audioUnitCreationWarmupPromise = m_audioUnitCreationWarmupPromise->whenSettled(RunLoop::main(), [callback = WTFMove(callback)] (auto&&) mutable {
```
this isn't necessary, no need to check if the promise is settled (ever).

The promise created above it either pending or settled; let `whenSettle` do the rest.

jyavenard replied:
plus it would make the callback run out of order (probably doesn't matter but still)

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,30 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
+    }
+
+    callback();
+}
```
so all this can be replaced by something like:
```
    if (m_audioUnitCreationWarmupPromise)
        return *m_audioUnitCreationWarmupPromise;

    m_audioUnitCreationWarmupPromise = invokeAsync(WorkQueue::create("CoreAudioSharedUnit AudioUnit creation").get(), [] {
        return NativePromise<CoreAudioSharedUnit::StoredAudioUnit, OSStatus>::createAndSettle(createAudioUnit(true));
    })->whenSettled(RunLoop::current(), [weakThis = WeakPtr { *this }](auto&& vpioUnitOrError) mutable {
        if (!weakThis)
            return GenericNonExclusivePromise::createAndReject();
        if (vpioUnitOrError.has_value())
            weakThis->setStoredVPIOUnit(WTFMove(vpioUnitOrError.value()));
        return GenericNonExclusivePromise::createAndResolve();
    });
    return *m_audioUnitCreationWarmupPromise;
```

with `m_audioUnitCreationWarmupPromise` being a RefPtr<GenericNonExclusive>

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
```diff
@@ -183,7 +192,10 @@ class CoreAudioSharedUnit final : public BaseAudioSharedUnit {
-    AudioUnit m_storedVPIOUnit { nullptr };
+#if PLATFORM(MAC)
+    StoredAudioUnit m_storedVPIOUnit { nullptr };
+    RefPtr<WTF::GenericNonExclusivePromise> m_audioUnitCreationWarmupPromise;
```
WTF:: isn't needed.

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,30 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
+{
+    if (!m_audioUnitCreationWarmupPromise) {
+        m_audioUnitCreationWarmupPromise = invokeAsync(WorkQueue::create("CoreAudioSharedUnit AudioUnit creation").get(), [weakThis = WeakPtr { *this }] () mutable {
+            callOnMainRunLoop([weakThis = WTFMove(weakThis), vpioUnitOrError = createAudioUnit(true)] () mutable {
```
I would still ditch `callOnMainRunLoop` here and use `whenSettled` on the promise returned by invokeAsync. with #20770 you can just do `return createAudioUnit(true)` in the lambda.
```
m_audioUnitCreationWarmupPromise = invokeAsync(WorkQueue::create("CoreAudioSharedUnit AudioUnit creation").get(), [] {
        return createAudioUnit(true);
    })->whenSettled(RunLoop::main(), [weakThis = WeakPtr { *this }](auto&& vpioUnitOrError) mutable {
        if (!weakThis)
            return GenericNonExclusivePromise::createAndReject();
        if (vpioUnitOrError.has_value())
            weakThis->setStoredVPIOUnit(WTFMove(vpioUnitOrError.value()));
        return GenericNonExclusivePromise::createAndResolve();
    });
```

jyavenard replied:
I should add that to me `m_audioUnitCreationWarmupPromise` is the promise that the AU has been created *and* is ready for use (that it got stored).

By dissociating the call to `setStoredVPIOUnit` having been run from the AU creation promise, it becomes much less clear that the AU is not just created but also usable by the CoreAudioSharedUnit users.

It is an implementation detail that the end result is the same in how the task are dispatched.

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,24 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
+        });
+    }
+
+    m_audioUnitCreationWarmupPromise->whenSettled(RunLoop::main(), [callback = WTFMove(callback)] (auto&&) mutable {
```
`auto&&` can be removed; the lambda given to whenSettled takes either an Expected or void

jyavenard commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -867,6 +867,11 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         // Defined -> operator for consistency in calling pattern.
         ThenCommand* operator->() { return this; }
 
+        operator RefPtr<PromiseType>()
```
Can we add this to the Producer class too?

Maybe those operator should be made pure virtual to ConvertibleToNative as you want all the classes that can be converted to NativePromise to behave the same way and provide the same sets of APIs.

jyavenard commented on Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
```diff
@@ -681,8 +682,24 @@ bool CoreAudioSharedUnit::migrateToNewDefaultDevice(const CaptureDevice& capture
+        })->whenSettled(RunLoop::main(), [weakThis = WeakPtr { *this }] (auto&& vpioUnitOrError) {
+            if (weakThis && vpioUnitOrError.has_value())
+                weakThis->setStoredVPIOUnit(WTFMove(vpioUnitOrError.value()));
+            return WTF::GenericNonExclusivePromise::createAndResolve();
```
WTF:: isn't needed.

jyavenard replied:
Should we handle errors ?

youennf replied:
No, this is prewarming. If it fails, we will just retry when actually starting to capture and fail capture if it fails again.

youennf committed:
[macOS] Create a VPIO unit in a background thread
https://bugs.webkit.org/show_bug.cgi?id=265017
rdar://118520768

Reviewed by Jean-Yves Avenard.

A VPIO Unit may take a few seconds to be created.
This might block the GPU process main thread, which might block the WebProcess main thread.
The impact might be a page freeze.
To prevent this, we create the VPIO unit in a background queue and asynchronously start capturing.

To do so, CoreAudioCaptureSource is implementing RealtimeMediaSource::whenReady, where we prewarm the audio unit.
The impact is to delay the resolution of the getUserMedia promise.

To improve the handling of AudioUnit, we introduce StoredAudioUnit to make sure we always dispose a created audio unit.
We then refactor the code to use move-only audio units.

Covered by existing tests in macOS with MockAudioSharedUnit.
Manually tested as well.

* Source/WTF/wtf/NativePromise.h:
* Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.h:
(WebCore::BaseAudioSharedUnit::prewarmAudioUnitCreation):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::whenReady):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.cpp:
(WebCore::CoreAudioSharedUnit::AudioUnitDeallocator::operator() const):
(WebCore::createAudioUnit):
(WebCore::CoreAudioSharedInternalUnit::create):
(WebCore::CoreAudioSharedInternalUnit::CoreAudioSharedInternalUnit):
(WebCore::CoreAudioSharedInternalUnit::~CoreAudioSharedInternalUnit):
(WebCore::CoreAudioSharedInternalUnit::initialize):
(WebCore::CoreAudioSharedInternalUnit::uninitialize):
(WebCore::CoreAudioSharedInternalUnit::start):
(WebCore::CoreAudioSharedInternalUnit::stop):
(WebCore::CoreAudioSharedInternalUnit::set):
(WebCore::CoreAudioSharedInternalUnit::get):
(WebCore::CoreAudioSharedInternalUnit::render):
(WebCore::CoreAudioSharedUnit::~CoreAudioSharedUnit):
(WebCore::CoreAudioSharedUnit::setStoredVPIOUnit):
(WebCore::CoreAudioSharedUnit::reconfigureAudioUnit):
(WebCore::CoreAudioSharedUnit::migrateToNewDefaultDevice):
(WebCore::CoreAudioSharedUnit::prewarmAudioUnitCreation):
(WebCore::CoreAudioSharedInternalUnit::storeVPIOUnitIfNeeded): Deleted.
* Source/WebCore/platform/mediastream/mac/CoreAudioSharedUnit.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):

Canonical link: https://commits.webkit.org/271050@main
